--- /usr/share/eselect/modules/profile.eselect	2011-09-18 23:35:04.000000000 +0400
+++ /home/ztime/profile.eselect	2012-01-08 22:13:05.000000000 +0400
@@ -17,6 +17,11 @@
 [[ -d ${NEW_CALCULATE_OVERLAY} ]] && 
 	CALCULATE_OVERLAY=${NEW_CALCULATE_OVERLAY} ||
 	CALCULATE_OVERLAY="/usr/local/portage/layman/calculate"
+# add ztime overlay to profiles
+NEW_ZTIME_OVERLAY="/var/lib/layman/ztime"
+[[ -d ${NEW_ZTIME_OVERLAY} ]] && 
+	ZTIME_OVERLAY=${NEW_ZTIME_OVERLAY} ||
+	ZTIME_OVERLAY="/usr/local/portage/layman/ztime"
 
 # get location of make.profile symlink
 get_symlink_location() {
@@ -52,6 +57,16 @@
 			echo ${p}
 		done
 	fi
+	# add support ztime overlay
+	if [[ -f "${ROOT}${ZTIME_OVERLAY}/profiles/profiles.desc" ]]
+	then
+		for p in $(sed -n -e \
+			"s|^${arch}[[:space:]]\+\([^[:space:]]\+\).*$|\1|p" \
+			"${ROOT}${ZTIME_OVERLAY}/profiles/profiles.desc")
+		do
+			echo ${p}
+		done
+	fi
 
 	for p in $(sed -n -e "s|^${arch}[[:space:]]\+\([^[:space:]]\+\).*$|\1|p" \
 		"${ROOT}${portdir}/profiles/profiles.desc")
@@ -106,6 +121,24 @@
 		if [[ ${arch} != ${parch} && -z ${force} ]] ; then
 			die -q "${target} is not a valid profile for ${arch}"
 		fi
+		# add support ztime overlay
+	elif [[ -n ${target} && -d ${ROOT}${ZTIME_OVERLAY}/profiles/${target} ]]
+	then
+		local arch parch
+
+		# if the profile was explicitly specified (rather than a number)
+		# double check and make sure it's valid
+		arch=$(arch)
+		[[ -z ${arch} && -z ${force} ]] && return 1
+
+		# do a reverse lookup and find the arch associated with ${target}
+		parch=$(sed -n -e \
+			"s|^\([[:alnum:]]\+\)[[:space:]].*${target}[[:space:]].*$|\1|p" \
+			"${ROOT}${ZTIME_OVERLAY}/profiles/profiles.desc")
+
+		if [[ ${arch} != ${parch} && -z ${force} ]] ; then
+			die -q "${target} is not a valid profile for ${arch}"
+		fi
 	fi
 
 	if [[ -z ${target} ]]; then
@@ -138,6 +171,17 @@
 		fi
 		ln -s "..${CALCULATE_OVERLAY}/profiles/${target}" \
 			"${EROOT}/etc/make.profile"
+	# add support ztime overlay
+	elif [[ -d ${ROOT}${ZTIME_OVERLAY}/profiles/${target} ]] ; then
+		# we must call remove_symlink() here instead of calling
+		# it from do_set(), since if the link is removed, we
+		# cannot determine $ARCH in find_targets()
+		if [[ -L ${MAKE_PROFILE} ]] ; then
+			remove_symlink \
+				|| die -q "Couldn't remove current make.profile symlink"
+		fi
+		ln -s "..${ZTIME_OVERLAY}/profiles/${target}" \
+			"${EROOT}/etc/make.profile"
 	else
 		die -q "Target \"$1\" doesn't appear to be valid!"
 	fi
@@ -160,6 +204,9 @@
 		# add support calculate overlay
 		calcprofiledir=$(canonicalise "${ROOT}${CALCULATE_OVERLAY}/profiles")
 		link=${link##${calcprofiledir}/}
+		# add support ztime overlay
+		ztimeprofiledir=$(canonicalise "${ROOT}${ZTIME_OVERLAY}/profiles")
+		link=${link##${ztimeprofiledir}/}
 
 		write_kv_list_entry "${link}" ""
 	else
@@ -187,6 +234,9 @@
 	# add support calculate overlay
 	calcprofiledir=$(canonicalise "${ROOT}${CALCULATE_OVERLAY}/profiles")
 	active=${active##${calcprofiledir}/}
+	# add support ztime overlay
+	ztimeprofiledir=$(canonicalise "${ROOT}${ZTIME_OVERLAY}/profiles")
+	active=${active##${ztimeprofiledir}/}
 
 	if [[ -n ${targets[@]} ]]; then
 		local i
